### PR TITLE
App ver. 0.2

### PR DESCRIPTION
**Pull request by iamandiradu**
This PR contains:
- App ver. 0.1 (BubbleMap of near coffee shops)
- App ver. 0.2 (WorldMap of all coffee shops & the ability to toggle between BubbleMap & WorldMap)
- Various configs (prettier, gitignore)
- Hotfix: maps loading before data is ready

### CODE REVIEW

**Review by alexandru-calinoiu — COMMENTED** [review]
Ideally you should include a small section on how to build and run, is there a minimal node version? Are there any env variables that need to be set?

It can be as minimal as https://github.com/amcharts/amcharts4#how-to-build

**Comment by iamandiradu** [comment]
Minimal NodeJS version: 10, required by gh-pages package. 
As far as I know, there are no env variables that need to be set.

I modified the README file to include the requested instructions.

File: src/components/app/app.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            .then((token) => {
+                setApiToken(token);
+                api.getData(token).then((data) => {
+                    setApiData(data);
+                });
+            })
+            .catch((error) => {
+                console.error(error);
+            });
+    }, []);
```
Comment (adrianbunea) [inline src/components/app/app.js]: Callback hell is real. Did you try to mark the functions defined by you as `async` and use `await` to flatten the code a bit?

File: src/components/app/app.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                });
+                coffeeShopsData.push({
+                    label: 'User',
+                    latitude: userCoordinates.latitude,
+                    longitude: userCoordinates.longitude,
+                    color: 'red',
+                    value: 1,
+                });
+            }
+            setProcessedApiData(coffeeShopsData);
```
Comment (adrianbunea) [inline src/components/app/app.js]: It looks like you're doing `forEach` on the apiData twice, once for calculating the distances, and another time for creating the data for the bubble/world map. Wouldn't it be better if you did this processing in a single loop? 

Also, you don't need to recalculate the distance of the shop when you create the tooltip, you already stored it the distance field of the coffee shop.

Comment (adrianbunea) [reply src/components/app/app.js]: The comparison of index with the correct number of items to be shown on the map took me a while to figure out. If you can toggle the map between the 2 views, then eventually you will need to calculate the coffee shop data for all 6 shops, wouldn't it make sense to just do all 6 at the start, and simply send 3 or 6 to be displayed based on which state you are in (sorted by distance of course)? This check makes me think the `coffeeShopsData` is recalculated from scratch every time the map is toggled, this doesn't look very efficient, can the model for the view be cached?

File: src/components/app/app.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                    ) : (
+                        <BubbleMap data={processedApiData} />
+                    )}
+                </div>
+            )}
+        </div>
+    );
+}
+
+export default App;
```
Comment (adrianbunea) [inline src/components/app/app.js:177]: Overall, the file seems to be doing a lot of things, maybe you can abstract the creation of `coffeeShopData` into a function, and since you do quite a lot of processing in each iteration to create each coffee shop item stored in the array, maybe you should also extract a function for it that is passed as the callback to the `forEach()`.

As a side note, functions that contain a single line of code can still be useful because they tell you what they do, while a single line of code might leave you guessing.

**Comment by iamandiradu** [comment]
Merged feature/xy-graph & feature/api-error-handling.
These 2 branches contain the updated XY Graph interface & improved API error handling.

File: src/utils/user-distance.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        var p = 0.017453292519943295; // Math.PI / 180
+        var c = Math.cos;
+        var a =
+            0.5 -
+            c((latitude - userCoordinates.latitude) * p) / 2 +
+            (c(userCoordinates.latitude * p) *
+                c(latitude * p) *
+                (1 - c((longitude - userCoordinates.longitude) * p))) /
+                2;
+        return Math.round((12742 * Math.asin(Math.sqrt(a)) + Number.EPSILON) * 100) / 100; // 2 * R; R = 6371 km
```
Comment (adrianbunea) [inline src/utils/user-distance.js:12]: Quite a few magical numbers in here, you point out what they mean by putting comments at the end, but you could just store these numbers in constants with descriptive names.
Storing `Math.cos` in a variable to make calling it less verbose might be a good idea, but `c` tells a lot less than `cos`. It is generally hard to know what single letter variables store, requiring the reader to have prior knowledge of what is being done here.
Finally, distance can be calculated with different methods, the formula used could be abstracted into a function with a descriptive name.

File: src/components/app/app.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        }
+        if (debouncedUserLongitude) {
+            setUserCoordinates((prevState) => ({
+                ...prevState,
+                longitude: debouncedUserLongitude,
+            }));
+        }
+    }, [debouncedUserLatitude, debouncedUserLongitude]);
+
+    return isLoading ? (
```
Comment (alexandru-calinoiu) [inline src/components/app/app.js:131]: Nice touch with loading :1st_place_medal: 

I've seen to many black screen staring back, leaving me wondering what to expect next.

File: src/components/app/app.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        <div className="spinnerWrapper">
```
Comment (alexandru-calinoiu) [inline src/components/app/app.js:132]: I consider that this view code part of this method and also to long ...

You can apply the same logic I've did above to figure out what blocks need to be extracted.

**Review by alexandru-calinoiu — COMMENTED** [review]
I like the visual implementation and the attention to user detail.

On the first run I am greated by this wonderfull error

image

Also this compile warning look rather easy to fix and don't leave a good impression

image

The lack of tests makes this a deal breaker for me.

File: src/utils/user-coordinates.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export default {
+    getUserCoordinates: () => {
```
Comment (biancamariacampean) [inline src/utils/user-coordinates.js]: I like that you extracted the functions in specific files :+1:

File: src/utils/user-distance.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+export function getUserDistance(userCoordinates, latitude, longitude) {
+    if (userCoordinates.latitude) {
+        var p = 0.017453292519943295; // Math.PI / 180
+        var c = Math.cos;
+        var a =
```
Comment (biancamariacampean) [inline src/utils/user-distance.js:5]: I would like to use other names for your variables

File: public/index.html [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
+        <!--
+      Notice the use of %PUBLIC_URL% in the tags above.
+      It will be replaced with the URL of the `public` folder during the build.
+      Only files inside the `public` folder can be referenced from the HTML.
+
+      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
+      work correctly both with client-side routing and a non-root public URL.
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
```
Comment (biancamariacampean) [inline public/index.html:23]: I would prefer if you delete these comments and leave only your comments (so your intervention can be seen). The code is hard to follow

File: src/components/app/app.js (hunk 1 above)
Comment (iamandiradu) [reply src/components/app/app.js]: React's useEffect issues a warning if I try to make it async. The recommended way to do it is by adding an async function and calling it before the closure of the useEffect hook. 
Initially, I had only 1 callback hell and it looked cleaner using promises. 
Will refactor this to flatten the code. :)

**Review by biancamariacampean — COMMENTED** [review]
Overall, your implementation looks good. I would like to see some tests to confirm that everything works correctly.

File: src/components/app/app.js (hunk 1 above)
Comment (iamandiradu) [reply src/components/app/app.js]: Callback hell refactored in this [commit](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/ecadfac36471840e4b47a7a8e48b749e0725d3e1).

File: src/components/app/app.js (hunk 2 above)
Comment (iamandiradu) [reply src/components/app/app.js]: Data processor refactored in this [commit](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/4253a9a775b766c0fcde147c4f55ac0339acc1f4).

File: src/components/app/app.js (hunk 3 above)
Comment (iamandiradu) [reply src/components/app/app.js:149]: I exported the data processor as advised. Check this [commit](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/893620ff4dee1a4900d4c7b1db7001956e5b1a60).

File: src/utils/user-distance.js (hunk 4 above)
Comment (iamandiradu) [reply src/utils/user-distance.js]: The method used to calculate the distances applies the Harversine formula. I modified the helper to be a bit more explicit.
The changes can be found in this [commit](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/0088fe8acd84b6520c64e46cc98b61d053bc3d00).

File: src/api/urls.js [hunk 10]
```diff
@@ -0,0 +1,5 @@
+const blueBottleAPI = 'https://blue-bottle-api-test.herokuapp.com/v1';
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/api/urls.js:1]: Usually, endpoints are also stored. In this case, the /tokens and /coffee_shops endpoints can also be added in this file.
I updated the file to also contain the 2 endpoints in use. This [commit](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/c105964110b62d2f18086eb1dad5ed1bf25755d1) should also fix the "error.response is undefined" encountered above.

File: src/api/api.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
+import axios from 'axios';
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/api/api.js:1]: The main reason for separating them was reusability and DRYness.
If the question is about using axios, then the reason is because axios has a wider browser support than fetch.

File: src/api/api.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+                headers: {
+                    Accept: 'application/json',
+                },
+            });
+            return response.data;
+        } catch (error) {
+            const errorStatus = error.response.status;
+            switch (errorStatus) {
+                case 401:
+                    console.log('401: Invalid token.');
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/api/api.js]: The token may not be valid yet or may have expired, thus returning a 401 error code. 
By default, that code means 'Unauthorized'.

File: src/utils/use-debounce.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import React, { useState, useEffect } from 'react';
+
+// Our hook
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/utils/use-debounce.js:3]: This hook is imported from @gragland [usehooks](https://github.com/gragland/usehooks) repo. Initially, I left them in the file to better understand how it works. I was planning on removing them and I got carried away.

I try to keep the in-line comments at a minimal, using them only where they are needed.
Also, to better organise and clarify code, I comment before functions (or elements) if they are not self explanatory.

File: src/utils/user-distance.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+export function getUserDistance(userCoordinates, latitude, longitude) {
+    if (userCoordinates.latitude) {
+        var p = 0.017453292519943295; // Math.PI / 180
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/utils/user-distance.js]: I modified the helper to be a bit more explicit.
The changes can be found in this [commit](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/0088fe8acd84b6520c64e46cc98b61d053bc3d00).

Using short names can make a difference in how big code files are, but doing this will defeat the purpose of minifiers. The usage of short names is against coding best practices, as it makes the code hard or impossible to read or understand.
Either if you use short or long naming, if a minifier is present, naming is irrelevant.

File: src/utils/use-debounce.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/utils/use-debounce.js]: This [commit](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/5012959956a897a786b55e5b6ba6b43ab4182398) contains a cleaner version of the useDebounce hook.

File: src/serviceWorker.js [hunk 15]
```diff
@@ -0,0 +1,141 @@
+// This optional code is used to register a service worker.
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/serviceWorker.js]: You guess is right. The serviceWorker is a leftover from the boilerplate. Because a PWA is not planned, I removed it [here](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/bbf996c1a63c24b7b1c9ba25e93dd8872b023f50).

File: src/components/app/app.js [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+    };
+
+    // Get API Data
+    useEffect(() => {
+        api.getToken()
+            .then((token) => {
+                setApiToken(token);
+                api.getData(token).then((data) => {
+                    setApiData(data);
+                });
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/components/app/app.js]: There is indeed. I refactored this part [here](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/ecadfac36471840e4b47a7a8e48b749e0725d3e1).

File: src/components/app/app.js [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+                                : coffeeShop.name,
+                            latitude: parseFloat(coffeeShop.x),
+                            longitude: parseFloat(coffeeShop.y),
+                            color: '#009ED9',
+                            customTooltip: `${getUserDistance(
+                                userCoordinates,
+                                coffeeShop.x,
+                                coffeeShop.y
+                            )} km`,
+                            value: 1,
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/components/app/app.js]: This value is required by AmCharts package to render the circles/bubbles with the same radius. Removing this property makes the circles' radius very small and a default value is not available.

File: public/index.html (hunk 9 above)
Comment (iamandiradu) [reply public/index.html]: Sorry about that. These are leftovers from the boilerplate app. I deleted these comments and other leftovers in this [commit](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/5d9f6761a5d2a2821b3a4c6c1cce2b686beb05f9).

File: src/components/app/app.js (hunk 6 above)
Comment (iamandiradu) [reply src/components/app/app.js]: I refactored app's rendering into separate functions. This should make the rendering part more readable and understandable. 
The changes can be seen [here](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/91939e0c1f724525141129e40e86e9dea873e09c).

**Comment by iamandiradu** [comment]
I merged feature/tests into this branch. The tests do not cover all the cases, but should be fit for a starting point.
TODO: 
      improve app error handling
      improve app tests

File: src/components/app/app.js (hunk 5 above)
Comment (iamandiradu) [reply src/components/app/app.js]: I should also add a type of loader when toggling between interfaces.

File: src/api/api.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+export default {
+    getToken: async () => {
+        try {
+            const response = await axios({
+                url: `${api.apiURL}/tokens`,
+                method: 'post',
+            });
+            return response.data.token;
+        } catch (error) {
+            console.error(error);
```
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/api/api.js]: These [changes](https://github.com/iamandiradu/agilefreaks_bluebottle/pull/5/commits/ecadfac36471840e4b47a7a8e48b749e0725d3e1) should allow the app to handle the errors better. 

Axios interceptors could also be used to improve the error handling part (refreshing the token is 401: Unauthorized, for example) and retry requests.

If an error reporting tool is added, the handling code would be adapted to also send logs to the given service (ex: Sentry) and capture errors.

File: src/utils/user-distance.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (adrianbunea) [reply src/utils/user-distance.js]: So, if using short names defeats the purpose of minifiers, and is against best practices, why are you still using one letter names here?

The problem's description said "Assume all coordinates lie on a plane", Haversine is overkill here.

File: src/components/app/app.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (adrianbunea) [reply src/components/app/app.js]: Isn't there a better way of conveying what 1 means other than by writting a comment?

File: src/utils/user-distance.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/utils/user-distance.js]: The initial code used to calculate the distance was a generic JS implementation of Haversine. Didn't think of renaming the variables as I didn't take an in-depth look over the formula. 

You're right about the requirement. Even tough it wasn't necessary, I wanted to show the accurate distance between the user and the coffee shops and went overboard with using the Haversine.

File: src/components/app/app.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (iamandiradu) [reply src/components/app/app.js]: There are 2 ways:
1. Naming a constant (ex: defaultGraphBubbleValue) and assigning it to the 'value' property.
2. Adding a default value in the forked library.

The bubbles take the smallest size allowed when no value is given. 
I tried finding a way of adding the default in AmCharts' docs, but didn't succeed. Leaving the field empty didn't help either.

File: public/index.html (hunk 9 above)
Comment (biancamariacampean) [reply public/index.html]: Nice :+1:

File: src/api/api.js [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+                Accept: 'application/json',
+            },
+            timeout: 1000,
+        });
+        return response.data;
+    } catch (error) {
+        if (error.response && error.response.status === 401) {
+            const token = await getToken();
+            return await getData(token);
+        }
```
Comment (alexandru-calinoiu) [inline src/api/api.js:34]: This is not the expected behaviour, the idea is when you get a 401 you should try to refresh the token (get it again in this case) and retry the last call once, if you get a 401 again then this means that the user session is no longer valid and you should let the user know.

File: src/components/app/app.js [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+                userCoordinates,
+                coffeeShopsShownOnMap,
+                coffeeShopNameSplitter
+            );
+
+            setProcessedApiData(coffeeShopsData);
+        }
+    }, [apiData, userCoordinates]);
+
+    // Set isLoading to false if all data is loaded
```
Comment (alexandru-calinoiu) [inline src/components/app/app.js:66]: One pitfall of comments is that they lie, is the same case here as this comment is not accurate, isLoading is false when there is an error also.

**Review by alexandru-calinoiu — COMMENTED** [review]
Looks good and it works as expected, the implementation is procedural and not very oop.

Left a couple of comments for your information they don't need to be addressed.

File: src/components/app/app.js (hunk 20 above)
Comment (iamandiradu) [reply src/components/app/app.js:66]: That's true. The comment does not cover the error part.

File: src/api/api.js (hunk 19 above)
Comment (iamandiradu) [reply src/api/api.js:34]: I noticed my mistake. Indeed, this is not the recommended way to handle expired sessions.
